lib/Parse/ParseDecl.cpp

```diff
@@ -4792,6 +4793,10 @@ bool Parser::isStartOfSwiftDecl(bool allowPoundIfAttributes) {
     return isStartOfSwiftDecl(allowPoundIfAttributes);
   }
 
+  // Macro expansion starts with a pound and a custom identifier.
+  if (Tok.is(tok::pound) && peekToken().is(tok::identifier))
+    return true;
```

Comment:
We hit some issues in the Swift Swift parser when we did this, and I ended up doing a bit more work in that parser to parse as macro expressions more often (https://github.com/apple/swift-syntax/pull/1296). With this change, do we still parse `#someMacro == 17` properly?

Reply:
`#someMacro == 17` broke a ton of things with parsing everything as a `MacroExpansionDecl`. I changed this back to paying things as an expr, and this simplified things quite a bit. The custom "post type-checking walks" are no longer needed.